python/xir/program.py

```diff
@@ -50,37 +50,63 @@ def __init__(self, name: str, params: Union[List, Dict], wires: Tuple, use_float
         self._params = params
         self._wires = wires
 
+        self._is_adjoint = False
+        self._ctrl_wires = None
+
         self._use_floats = use_floats
 
     def __str__(self):
+        """Serialized string representation of a Statement"""
         if isinstance(self.params, dict):
             params = [f"{k}: {v}" for k, v in self.params.items()]
         else:
             params = [str(p) for p in self.params]
         params_str = ", ".join(params)
+        if params_str != "":
+            params_str = "(" + params_str + ")"
 
         wires = ", ".join([str(w) for w in self.wires])
 
-        if params_str == "":
-            return f"{self.name} | [{wires}]"
-        return f"{self.name}({params_str}) | [{wires}]"
+        modifier_str = ""
+        if self.ctrl_wires is not None:
+            ctrl_wires = ", ".join([str(w) for w in self.ctrl_wires])
+            modifier_str = f"ctrl[{ctrl_wires}] "
+        if self.is_adjoint:
+            modifier_str = "adjoint " + modifier_str
+
+        return f"{modifier_str}{self.name}{params_str} | [{wires}]"
 
     @property
     def name(self) -> str:
+        """Returns the name of the gate statement"""
         return self._name
 
     @property
     def params(self) -> Union[List, Dict]:
+        """Returns the parameters of the gate statement"""
         if self.use_floats:
             return get_floats(self._params)
         return self._params
 
     @property
     def wires(self) -> Tuple:
+        """Returns the wires that the gate is applied to"""
         return self._wires
 
+    @property
+    def is_adjoint(self) -> bool:
+        """Whether the statement is an adjoint gate"""
+        return self._is_adjoint
+
+    @property
+    def ctrl_wires(self) -> Optional[tuple]:
```

Comment:
Two things:
1. Python 3.7 does not support lowercase generic types so we'll need to use `Tuple` instead.
2. We should also specify the type of the items in the tuple too.
```suggestion
    def ctrl_wires(self) -> Optional[Tuple[int, ...]]:
```
Alternatively, we can encapsulate the container of the wires using an iterator:
```suggestion
    def ctrl_wires(self) -> Optional[Iterator[int]]:
```
--- 

Note that point (2) also applies to the other functions in this file.  As a user of the XIR, I usually find myself deriving the types of the items manually if they are not explicitly specified in a function signature.

Reply:
Wires could be both integers or strings (names) here though. Would you suggest having `Optional[Iterator[Union[int, str]]]`, or only `Optional[Tuple]` or perhaps even defining a new type wires for this?

Reply:
> Wires could be both integers or strings (names) here though. 

Oh yeah, good point!

> Wires could be both integers or strings (names) here though. Would you suggest having `Optional[Iterator[Union[int, str]]]`, or only `Optional[Tuple]` or perhaps even defining a new type wires for this?

Good question! If it were up to me, I would create an alias like you suggested
```python
Wire = Union[int, str]
```
and write the function signature as follows:
```python
def ctrl_wires(self) -> Optional[Iterator[Wire]]:
```
In general, I'm in favour of taking the time to explicitly define each type in a function signature.  In this case, knowing the return type is useful since the client code may need to deal with `str` and `int` wire IDs separately.

Reply:
What's the motivation behind using `Tuple` instead of `Iterator` here?

Reply:
Just that wires are always kept as `Tuple`s internally, so I thought that'd be more correct. Would you prefer using `Iterator` instead?

Reply:
Ideally, we would return either a [`Sequence`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence) or an [`Iterator`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterator) (in hindsight, I may actually prefer the former) so that external code does that rely on the fact that we use a tuple to represent wires internally.